app/src/main/java/org/projectbuendia/client/sync/controllers/IncrementalSyncPhaseRunnable.java

```diff
@@ -0,0 +1,190 @@
+
+import org.projectbuendia.client.App;
+import org.projectbuendia.client.json.IncrementalSyncResponse;
+import org.projectbuendia.client.json.Serializers;
+import org.projectbuendia.client.net.Common;
+import org.projectbuendia.client.net.GsonRequest;
+import org.projectbuendia.client.net.OpenMrsConnectionDetails;
+import org.projectbuendia.client.providers.Contracts;
+import org.projectbuendia.client.sync.SyncAdapter;
+import org.projectbuendia.client.utils.Logger;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import static org.projectbuendia.client.net.OpenMrsServer.wrapErrorListener;
+
+/**
+ * Implements the basic logic for an incremental sync phase.
+ * <p>
+ * To implement an incremental sync phase, create a subclass, supply the appropriate arguments to
+ * {@link IncrementalSyncPhaseRunnable}'s constructor from the subclasses' public, no-arg
+ * constructor, and then implement the {@link #getUpdateOps(Object[], SyncResult)} method.
+ * <p>
+ * Note: you may also wish to undertake an action at the start and end of the sync phase - hooks are
+ * provided for this. See {@link #beforeSyncStarted(ContentResolver, SyncResult,
+ * ContentProviderClient)} and {@link #afterSyncFinished(ContentResolver, SyncResult,
+ * ContentProviderClient)}.
+ */
+public abstract class IncrementalSyncPhaseRunnable<T> implements SyncPhaseRunnable {
+
+    private static final Logger LOG = Logger.create();
+
+    private final String resourceType;
+    private final Contracts.Table dbTable;
+    private final Class<T> clazz;
+
+    /**
+     * Instantiate a new IncrementalSyncPhaseRunnable. This is designed to be called from a no-arg
+     * constructor of subclasses.
+     *
+     * @param resourceType the type of resource to be fetched. This string is appended to
+     *                     {@link OpenMrsConnectionDetails#getBuendiaApiUrl()} to determine the HTTP
+     *                     endpoint to request from.
+     * @param dbTable      the database table to store fetched data in. Also used as a key against
+     *                     which sync tokens from the server will be stored.
+     * @param clazz        the {@link Class} object corresponding to the generic type {@code T}.
+     */
+    protected IncrementalSyncPhaseRunnable(
+            String resourceType,
+            Contracts.Table dbTable,
+            Class<T> clazz) {
+        this.resourceType = resourceType;
+        this.dbTable = dbTable;
+        this.clazz = clazz;
+    }
+
+    @Override
+    public final void sync(ContentResolver contentResolver, SyncResult syncResult,
+        ContentProviderClient providerClient) throws Throwable {
+
+        boolean okToProceedSync = beforeSyncStarted(contentResolver, syncResult, providerClient);
+
+        if(!okToProceedSync) {
+            LOG.w("Skipping synchronization for %s", this.getClass().getName());
```

Comment:
Use `getClass().getSimpleName()` instead for this, `getName()` is the fully-qualified verbose name.